core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -3,156 +3,406 @@
+                    hook
+                  )
+                else
+                  strategy match {
+                    case BackPressure =>
+                      IO.now(true) -> Surplus(
+                        IQueue.empty[A].enqueue(addToQueue.toList),
+                        IQueue.empty.enqueue(surplusValues -> p),
+                        hook
+                      )
+                    case Sliding =>
+                      val notTaken = addToQueue ++ surplusValues
+                      val toQueue  = capacity.fold(notTaken)(notTaken.takeRight)
+                      p.complete(false) -> Surplus(
+                        IQueue.empty.enqueue(toQueue.toList),
+                        IQueue.empty,
+                        hook
+                      )
+                    case Dropping =>
+                      p.complete(false) -> Surplus(
+                        IQueue.empty.enqueue(addToQueue.toList),
+                        IQueue.empty,
+                        hook
+                      )
+                  }
+
+              completeTakers *> complete -> surplus
+            } else
+              completeTakers *> p.complete(true) -> Deficit(deficitValues, hook)
+        }
+
+      case (p, Surplus(values, putters, hook)) =>
+        val (addToQueue, surplusValues) =
+          capacity.fold((as, Iterable.empty[A]))(c => as.splitAt(c - values.size))
+        if (surplusValues.isEmpty)
+          (p.complete(true), Surplus(values.enqueue(addToQueue.toList), putters, hook))
+        else {
+          strategy match {
+            case BackPressure =>
+              (
+                IO.now(true),
+                Surplus(values.enqueue(addToQueue.toList), putters.enqueue(surplusValues -> p), hook)
+              )
+            case Sliding =>
+              (
+                p.complete(true),
+                capacity.fold(
+                  Surplus(values.enqueue(as.toList), putters, hook)
+                ) { c =>
+                  Surplus(
+                    values
+                      .takeRight(c - as.size)
+                      .enqueue(as.takeRight(c).toList),
+                    putters,
+                    hook
+                  )
+                }
+              )
+            case Dropping =>
+              (
+                p.complete(false),
+                Surplus(values.enqueue(addToQueue.toList), putters, hook)
+              )
+          }
+        }
+
+      case (p, state @ Shutdown) => (p.interrupt, state)
+    }
+
+    val release: (Boolean, Promise[Nothing, Boolean]) => IO[Nothing, Unit] = {
+      case (_, p) => p.poll.void <> removePutter(p)
+    }
+
+    Promise.bracket[Nothing, State[A], Boolean, Boolean](ref)(acquire)(release)
   }
+
+  /**
+   * Adds a shutdown hook that will be executed when `shutdown` is called.
+   * If the queue is already shutdown, the hook will be executed immediately.
+   */
+  final def onShutdown(io: IO[Nothing, Unit]): IO[Nothing, Unit] =
```

Comment:
Since this callback will be called a single time, we can slightly change the interface to "await"-style:

```scala
final def awaitShutdown: IO[Nothing, Unit]
```

You could implement with a promise, or even just wrap the existing implementation:

```scala
final def awaitShutdown: IO[Nothing, Unit] = 
  Promise.make[Nothing, Unit].flatMap(promise => {
    val hook = promise.complete(())
    IO.flatten(ref.modify {
      case Deficit(takers, hook)         => IO.unit -> Deficit(takers, hook *> io)
      case Surplus(queue, putters, hook) => IO.unit -> Surplus(queue, putters, hook *> io)
      case state @ Shutdown              => io      -> state
    }) *> promise.get
  })
```

This is a cleaner API.

Reply:
What is the difference? when we wrapped it into Promise we will complete it ? where we'll use the Promise `hook`?
And `awaitShutdown(io)` could be called many times (as I remember, like in the test: `e54` ) and we will run it even if the Queue is already shut down.

Reply:
See the signature above for the difference: `awaitShutdown` does not take any parameters. Instead, it returns an `IO` that is initially suspended, but resumes when the shutdown is complete. You don't need to change the internal representation, because you can use the `hook` to complete the `promise`, as shown above in the code snippet.

We have to maintain the same guarantee: `awaitShutdown` can be called many times. But for already an shutdown queue, it will resume right away.

Reply:
ah because I saw the `io` in `Shutdown` case, so in that case, we complete the promise (we call `hook`) right?

Reply:
Did I understand that wrongly..? 
now when I read again, onShutdown will stays and there we'll call awaitShutdown, right? ..